src/catalog/src/durable/impls/persist.rs

```diff
@@ -126,103 +123,80 @@ impl PersistHandle {
-                        }
-                    )
-                })
-                .expect("initialized catalog must have an epoch");
-            match initial_snapshot.remove(epoch_idx) {
-                StateUpdate {
-                    kind: StateUpdateKind::Epoch(epoch),
-                    diff,
-                    ..
-                } => {
-                    soft_assert_eq!(diff, 1);
-                    Some(epoch)
-                }
-                _ => unreachable!("checked above"),
-            }
+        // Fence out previous catalogs.
+        let mut fence_updates = Vec::with_capacity(2);
+        let prev_epoch = if upper != Timestamp::minimum() {
+            let as_of = self.as_of(upper);
+            let prev_epoch = self.get_epoch(as_of).await;
+            fence_updates.push(StateUpdate {
+                kind: StateUpdateKind::Epoch(prev_epoch),
+                ts: upper,
+                diff: -1,
+            });
+            Some(prev_epoch.get())
         } else {
             None
         };
-        let mut current_epoch = prev_epoch.map(|epoch| epoch.get()).unwrap_or(1);
-        // Note only writable catalogs attempt to increment the epoch.
+        let mut current_epoch = prev_epoch.unwrap_or(1);
+        // Only writable catalogs attempt to increment the epoch.
         if matches!(mode, Mode::Writable) {
             current_epoch = current_epoch + 1;
         }
         let current_epoch = Epoch::new(current_epoch).expect("known to be non-zero");
+        fence_updates.push(StateUpdate {
+            kind: StateUpdateKind::Epoch(current_epoch),
+            ts: upper,
+            diff: 1,
+        });
+        debug!(
+            ?upper,
+            ?prev_epoch,
+            ?current_epoch,
+            "fencing previous catalogs"
+        );
+        if matches!(mode, Mode::Writable) {
+            let next_upper = upper.step_forward();
+            self.compare_and_append(fence_updates, upper, next_upper)
+                .await?;
+        }
 
+        let (is_initialized, upper) = self.is_initialized_inner().await;
+        if !matches!(mode, Mode::Writable) && !is_initialized {
+            return Err(CatalogError::Durable(DurableCatalogError::NotWritable(
+                format!("catalog tables do not exist; will not create in {mode:?} mode"),
+            )));
+        }
+        soft_assert_ne!(upper, Timestamp::minimum());
+        let restart_as_of = self.as_of(upper);
+        let user_version = self.get_user_version(restart_as_of).await;
         debug!(
             ?is_initialized,
             ?upper,
+            ?restart_as_of,
             ?user_version,
-            ?prev_epoch,
-            ?current_epoch,
-            "open inner"
+            "initializing catalog state"
         );
 
+        let subscribe = self
+            .read_handle
+            .subscribe(Antichain::from_elem(restart_as_of))
+            .await
+            .expect("invalid usage");
```

Comment:
Whats the motivation behind using a `Subscribe` instead of a listen?

Reply:
It's just a tiny bit cleaner. A `Subscribe` is a snapshot at some time + a `Listen` starting at all times greater. So instead of

1. Get a snapshot at `t`.
2. Apply snapshot in memory.
3. Start listening at `t` and apply all listen events in memory.

We can:

1. Start subscribing at `t` and apply all subscribe events in memory.